hudi-spark/src/main/java/org/apache/hudi/QuickstartUtils.java

```diff
@@ -0,0 +1,319 @@
+     */
+    public static TestRawTripPayload generateRandomValue(HoodieKey key, String commitTime) throws IOException {
+      GenericRecord rec = generateGenericRecord(key.getRecordKey(), "rider-" + commitTime, "driver-" + commitTime, 0.0);
+      return new TestRawTripPayload(rec.toString(), key.getRecordKey(), key.getPartitionPath(), TRIP_EXAMPLE_SCHEMA);
+    }
+
+    /**
+     * Generates new inserts, uniformly across the partition paths above. It also updates the list of existing keys.
+     */
+    public Stream<HoodieRecord> generateInsertsStream(String commitTime, Integer n) {
+      int currSize = getNumExistingKeys();
+
+      return IntStream.range(0, n).boxed().map(i -> {
+        String partitionPath = partitionPaths[rand.nextInt(partitionPaths.length)];
+        HoodieKey key = new HoodieKey(UUID.randomUUID().toString(), partitionPath);
+        KeyPartition kp = new KeyPartition();
+        kp.key = key;
+        kp.partitionPath = partitionPath;
+        existingKeys.put(currSize + i, kp);
+        numExistingKeys++;
+        try {
+          return new HoodieRecord(key, generateRandomValue(key, commitTime));
+        } catch (IOException e) {
+          throw new HoodieIOException(e.getMessage(), e);
+        }
+      });
+    }
+
+    /**
+     * Generates new inserts, uniformly across the partition paths above. It also updates the list of existing keys.
+     */
+    public List<HoodieRecord> generateInserts(Integer n) throws IOException {
+      String commitTime = generateRandomCommitMsg();
+      return generateInsertsStream(commitTime, n).collect(Collectors.toList());
+    }
+
+    public HoodieRecord generateUpdateRecord(HoodieKey key, String commitTime) throws IOException {
+      return new HoodieRecord(key, generateRandomValue(key, commitTime));
+    }
+
+    /**
+     * Generates new updates, randomly distributed across the keys above. There can be duplicates within the returned
+     * list
+     *
+     * @param n Number of updates (including dups)
+     * @return list of hoodie record updates
+     */
+    public List<HoodieRecord> generateUpdates(Integer n) throws IOException {
+      String commitTime = generateRandomCommitMsg();
+      List<HoodieRecord> updates = new ArrayList<>();
+      for (int i = 0; i < n; i++) {
+        KeyPartition kp = existingKeys.get(rand.nextInt(numExistingKeys - 1));
+        HoodieRecord record = generateUpdateRecord(kp.key, commitTime);
+        updates.add(record);
+      }
+      return updates;
+    }
+
+    public static class KeyPartition implements Serializable {
+      HoodieKey key;
+      String partitionPath;
+    }
+
+    public void close() {
+      existingKeys.clear();
+    }
+  }
+
+  /**
+   * Example row change event based on some example data used by quickstart guide.
+   */
+  static class TestRawTripPayload implements HoodieRecordPayload<TestRawTripPayload> {
```

Comment:
I think this class is implementing  HoodieRecordPayload  directly due to historical reasons. There is a generic OverwriteWithLatestAvroPayload which you can reuse.   OverwriteWithLatestAvroPayload keeps it as Avro record and avoids json. 
If you need to showcase soft deletes, you can have this class extend OverwriteWithLatestAvroPayload, have "is_deleted" field part of the schema and only override getInsertValue() to check if this flag is enabled after decoding and return accordingly.

That ways, users who look at this class do not feel like they have to write a lot of code.

Reply:
This is neat. Thanks for suggesting this. I am planning to showcase delete in next iteration. So for now, I ll just reuse OverwriteWithLatestAvroPayload.